p2p/memp2p/memp2p_network.go

```diff
@@ -0,0 +1,174 @@
+// internal map of peers.
+func NewMemP2PNetwork() (*MemP2PNetwork, error) {
+	var peerIDs []p2p.PeerID
+	var messages []p2p.MessageP2P
+
+	network := MemP2PNetwork{
+		mutex:           sync.RWMutex{},
+		messageLogMutex: sync.RWMutex{},
+		peerIDs:         peerIDs,
+		peers:           make(map[p2p.PeerID]*MemP2PMessenger),
+		LogMessages:     false,
+		Messages:        messages,
+	}
+
+	return &network, nil
+}
+
+// ListAddresses provides the addresses of the known peers.
+func (network *MemP2PNetwork) ListAddresses() []string {
+	network.mutex.RLock()
+	addresses := make([]string, len(network.peerIDs))
+	i := 0
+	for _, peerID := range network.peerIDs {
+		addresses[i] = fmt.Sprintf("/memp2p/%s", peerID)
+		i++
+	}
+	network.mutex.RUnlock()
+	return addresses
+}
+
+// ListAddressesExceptOne provides the addresses of the known peers, except a specified one.
+func (network *MemP2PNetwork) ListAddressesExceptOne(peerIDToExclude p2p.PeerID) []string {
+	network.mutex.RLock()
+	resultingLength := len(network.peerIDs) - 1
+	if resultingLength <= 0 {
+		network.mutex.RUnlock()
+		return []string{}
+	}
+	addresses := make([]string, resultingLength)
+	k := 0
+	for _, peerID := range network.peerIDs {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		addresses[k] = fmt.Sprintf("/memp2p/%s", peerID)
+		k++
+	}
+	network.mutex.RUnlock()
+	return addresses
+}
+
+// Peers provides a copy of its internal map of peers
+func (network *MemP2PNetwork) Peers() map[p2p.PeerID]*MemP2PMessenger {
+	network.mutex.RLock()
+	peersCopy := make(map[p2p.PeerID]*MemP2PMessenger)
+	for peerID, peer := range network.peers {
+		peersCopy[peerID] = peer
+	}
+	network.mutex.RUnlock()
+	return peersCopy
+}
+
+// PeersExceptOne provides a copy of its internal map of peers, excluding a specific peer.
+func (network *MemP2PNetwork) PeersExceptOne(peerIDToExclude p2p.PeerID) map[p2p.PeerID]*MemP2PMessenger {
+	network.mutex.RLock()
+	peersCopy := make(map[p2p.PeerID]*MemP2PMessenger)
+	for peerID, peer := range network.peers {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		peersCopy[peerID] = peer
+	}
+	network.mutex.RUnlock()
+	return peersCopy
+}
+
+// PeerIDs provides a copy of its internal slice of peerIDs
+func (network *MemP2PNetwork) PeerIDs() []p2p.PeerID {
+	network.mutex.RLock()
+	peerIDsCopy := make([]p2p.PeerID, len(network.peerIDs))
+	for i, peer := range network.peerIDs {
+		peerIDsCopy[i] = peer
+	}
+	network.mutex.RUnlock()
+	return peerIDsCopy
+}
+
+//PeerIDsExceptOne provides a copy of its internal slice of peerIDs, excluding a specific peer.
+func (network *MemP2PNetwork) PeerIDsExceptOne(peerIDToExclude p2p.PeerID) []p2p.PeerID {
+	network.mutex.RLock()
+	resultingLength := len(network.peerIDs) - 1
+	if resultingLength <= 0 {
+		network.mutex.RUnlock()
+		return []p2p.PeerID{}
+	}
+
+	peerIDsCopy := make([]p2p.PeerID, resultingLength)
+	k := 0
+	for _, peerID := range network.peerIDs {
```

Comment:
same here, declare k inside the for-range

Reply:
Not here, `k` must be incremented separately, because the `peersIDsCopy` slice will be shorter by one specific element.

Reply:
Right, missed the `continue` statement.

Reply:
Right, missed the `continue` statement